base/server/python/pki/server/cli/migrate.py

```diff
@@ -601,3 +602,45 @@ def create_link(self, instance, source, dest):
 
         os.symlink(source, dest)
         os.lchown(dest, instance.uid, instance.gid)
+
+    def migrate_service(self, instance):
+        self.migrate_service_java_home(instance)
+
+    def migrate_service_java_home(self, instance):
+        # When JAVA_HOME in the Tomcat service config differs from the
+        # value in /usr/share/pki/etc/pki.conf, update the value in
+        # the service config.
+
+        if "JAVA_HOME" not in os.environ or not os.environ["JAVA_HOME"]:
+            logger.debug("Refusing to migrate JAVA_HOME with missing environment variable")
+
+        java_home = os.environ["JAVA_HOME"]
+
+        # Update in /etc/sysconfig/<instance>
+        result = self.update_java_home_in_config(instance.service_conf, java_home)
+        self.write_config(instance.service_conf, result)
+
+        # Update in /etc/pki/<instance>/tomcat.conf
+        result = self.update_java_home_in_config(instance.tomcat_conf, java_home)
+        self.write_config(instance.tomcat_conf, result)
+
+    def update_java_home_in_config(self, path, new_value):
+        result = []
+
+        with open(path, 'r') as conf_fp:
+            for line in conf_fp:
+                target="JAVA_HOME="
+                if target not in line:
```

Comment:
Are we trying to match just `JAVA_HOME` or some other values as well? If just `JAVA_HOME`, I think we should match only at the beginning of the line. If multiple values, I think we should split the line by the delimiter then compare it against a list of acceptable values.

Reply:
I don't think that's sufficient. What if the user commented out `JAVA_HOME`? Tomcat will start fine with a commented out `JAVA_HOME` and will attempt to discover the default system `JAVA_HOME` path. In our case, this should work fine (because we're building with the default).

However, if they later uncomment it out when it had say, JDK 8 from F32 but they've since upgraded to F33/JDK11... Now they're not using the JDK11 we built with.

I think, I'm fine leaving `JAVA_HOME` commented out if it is already (hence, preserving `prefix`) -- but I'm not fine ignoring commented `JAVA_HOME` lines.

Would it be better to do more parsing? Perhaps:
 - Check for matched quotes
 - Only replace the path component
 - Leave any trailing data
 - Validate that there isn't another letter or underscore in front of `JAVA_HOME`?

The issue is that the initial comment prefix could be any number of `#` and ` `s combined, so just checking for one might not get everyone's config. Perhaps that's fine?

Or yes, we can leave it to only stock `^JAVA_HOME="/path/...."` lines.

Reply:
I guess, if they ever uncomment `JAVA_HOME`, we'd replace it then, hm.

Reply:
I think in general it's fair to say that comments are outside the scope of our upgrade scripts since it's free-form and we don't know the admin's intention (e.g. whether they want to preserve the original value), and yes, the only one that matters is the uncommented one.